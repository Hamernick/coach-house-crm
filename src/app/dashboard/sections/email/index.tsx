export default function EmailSection() {
  return (
    <div>
      <h1>Email</h1>
      <p>This is the Email section.</p>
    </div>
  );
}
