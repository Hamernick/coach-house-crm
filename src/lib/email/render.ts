export type EmailBlock = {
  id: string
  type: "heading" | "paragraph"
  content: string
}

function escapeHtml(str: string) {
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#039;")
}

export function renderBlocksToHtml(blocks: EmailBlock[]): string {
  return blocks
    .map((block) => {
      const content = escapeHtml(block.content)
      switch (block.type) {
        case "heading":
          return `<h1>${content}</h1>`
        case "paragraph":
          return `<p>${content}</p>`
        default:
          return ""
      }
    })
    .join("\n")
}

export { renderBlocksToHtml as renderEmail }
