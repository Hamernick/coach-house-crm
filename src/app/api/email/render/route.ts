import { NextRequest, NextResponse } from "next/server";
import { renderEmail, EmailBlock } from "@/lib/email/render";
import { jsonError } from "@/lib/api";

function applyVariables(html: string, variables: Record<string, string>): string {
  return Object.entries(variables || {}).reduce(
    (acc, [key, value]) =>
      acc.replace(new RegExp(`{{\\s*${key}\\s*}}`, "g"), String(value)),
    html,
  );
}

export async function POST(req: NextRequest) {
  try {
    const { content_json, variables } = await req.json();
    const blocks: EmailBlock[] =
      typeof content_json === "string" ? JSON.parse(content_json) : content_json;
    let html = renderEmail(blocks);
    if (variables && typeof variables === "object") {
      html = applyVariables(html, variables as Record<string, string>);
    }
    return NextResponse.json({ html });
  } catch (err) {
    return jsonError(400, "Invalid content_json");
  }
}