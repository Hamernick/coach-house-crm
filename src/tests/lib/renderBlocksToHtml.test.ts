import { describe, it, expect } from 'vitest'
import { renderBlocksToHtml, EmailBlock } from '@/lib/email/render'

describe('renderBlocksToHtml', () => {
  it('renders blocks to html snapshot', () => {
    const blocks: EmailBlock[] = [
      { id: '1', type: 'heading', content: 'Hello <World>' },
      { id: '2', type: 'paragraph', content: 'Paragraph & more' },
    ]
    const html = renderBlocksToHtml(blocks)
    expect(html).toMatchSnapshot()
  })
})
