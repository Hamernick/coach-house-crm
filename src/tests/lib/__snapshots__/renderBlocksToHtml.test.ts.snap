exports[`renderBlocksToHtml renders blocks to html snapshot 1`] = `
<h1>Hello &lt;World&gt;</h1>
<p>Paragraph &amp; more</p>
`
